pkg/scheduler/framework/plugins/volumebinding/volume_binding.go

```diff
@@ -98,9 +120,137 @@ func (pl *VolumeBinding) Filter(ctx context.Context, cs *framework.CycleState, p
 	return nil
 }
 
-// New initializes a new plugin with volume binder and returns it.
-func New(_ *runtime.Unknown, fh framework.FrameworkHandle) (framework.Plugin, error) {
+// Reserve is called by the scheduling framework when the scheduler cache is
+// updated.
+func (pl *VolumeBinding) Reserve(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
+	allBound, err := pl.binder.Binder.AssumePodVolumes(pod, nodeName)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	cs.Write(allBoundStateKey, allBoundStateData(allBound))
+	return nil
+}
+
+// PreBind is called before binding a pod. All prebind plugins must return
+// success or the pod will be rejected and won't be sent for binding.
+//
+// PreBind will make the API update with the assumed bindings and wait until
+// the PV controller has completely finished the binding operation.
+//
+// If binding errors, times out or gets undone, then an error will be returned to
+// retry scheduling.
+func (pl *VolumeBinding) PreBind(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
+	state, err := cs.Read(allBoundStateKey)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	allBound, ok := state.(allBoundStateData)
+	if !ok {
+		return framework.NewStatus(framework.Error, "unable to convert state into allBoundStateData")
+	}
+	if allBound {
+		// no need to bind volumes
+		return nil
+	}
+	klog.V(5).Infof("Trying to bind volumes for pod \"%v/%v\"", pod.Namespace, pod.Name)
+	err = pl.binder.Binder.BindPodVolumes(pod)
+	if err != nil {
+		klog.V(1).Infof("Failed to bind volumes for pod \"%v/%v\": %v", pod.Namespace, pod.Name, err)
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	klog.V(5).Infof("Success binding volumes for pod \"%v/%v\"", pod.Namespace, pod.Name)
+	return nil
+}
+
+// Unreserve is called by the scheduling framework when a reserved pod was
+// rejected in a later phase.
+// TODO(cofyc) Revert assumed PV/PVC cache, see http://issues.k8s.io/82934#issuecomment-538269188
+func (pl *VolumeBinding) Unreserve(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) {
+	pl.binder.DeletePodBindings(pod)
```

Comment:
This logic along with PostBind() seems to be newly introduced? Could you explain why it's needed? As in the before, `DeletePodBindings()` is only called in deletePodFromSchedulingQueue().

Reply:
in `deletePodFromSchedulingQueue()`, pod is deleted when 1) deleted from kube-apiserver 2) scheduled onto a node. I didn't try to filter out scheduled pods in this plugin. So I need to clear pod bindings in `PostBind` after the pod is scheduled. Otherwise, the cache of scheduled pods will not be cleared.

`DeletePodBindings()` in `Unreserve()` is newly introduced because pod bindings cache is not needed when the pod is unreserved and put back into the queue. This is optional because the pods in the queue will be scheduled again soon and pod binding cache will be recreated.

Reply:
If the above is true, we should rename the function to something like `ClearPodCache`

Reply:
> Can you please clarify if this is true: a pod's entry in the bindingDecisions map is only relevant while the pod is being scheduled, and we previously didn't clear those decisions in case of errors because FindPodVolumes called in Filter will simply override them when we attempt to schedule the pod again.

yes, it's only relevant while the pod is being scheduled. previously, no shared `CycleState` between `Filter/Assume/Bind` phases, so we save pod volume binding decisions in this map. we didn't clear those decisions in cases of errors because the scheduler may put the pod into the active queue while the pod is in asynchronous binding and will fail previous volume binding operation if it clears volume binding decisions. 

I plan to refactor this to leverage `framework.CycleState`. Event handler registered in `New` can be removed too.

> Also, can you please confirm that a pod's entry in the bindingDecisions map doesn't impact the scheduling decision of other pods (as in the cache isn't global and it doesn't store reservation decisions that get checked across pods).

I confirm that it does not impact the scheduling decision of other pods.